File: backend/app/schemas/parent_portal.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.guardian import PreferredContactChannel


class ParentProfileResponse(BaseModel):
    guardian_id: uuid.UUID
    name: str
    email: str
    phone: str | None
    address: str | None
    occupation: str | None
    preferred_contact_channel: PreferredContactChannel


class ParentProfileUpdate(BaseModel):
    address: str | None = Field(default=None, max_length=500)
    occupation: str | None = Field(default=None, max_length=255)
    preferred_contact_channel: PreferredContactChannel | None = None

    model_config = {"extra": "forbid"}


class ParentChildPermissions(BaseModel):
    attendance: bool
    results: bool
    assignments: bool
    finance: bool
    messaging: bool
    pickup: bool


class ParentChildResponse(BaseModel):
    id: uuid.UUID
    admission_number: str
    first_name: str
    middle_name: str | None
    last_name: str
    date_of_birth: date
    gender: str
    photo_url: str | None
    class_id: uuid.UUID | None
    class_name: str | None
    class_level: str | None
    academic_session: str | None
    permissions: ParentChildPermissions


class PaginatedParentChildren(BaseModel):
    items: list[ParentChildResponse]
    total: int
    page: int
    per_page: int
    total_pages: int


class ParentDashboardResponse(BaseModel):
    children: list[ParentChildResponse]
    child_count: int
    attendance_records: int
    present_records: int
    attendance_rate: float
    approved_results: int
    upcoming_assignments: int


class ParentAttendanceItem(BaseModel):
    id: uuid.UUID
    date: date
    status: str
    note: str | None


class ParentResultItem(BaseModel):
    id: uuid.UUID
    subject: str
    academic_session: str
    term: str
    ca_score: float | None
    exam_score: float | None
    total_score: float | None
    grade: str | None
    teacher_comment: str | None


class ParentAssignmentItem(BaseModel):
    id: uuid.UUID
    title: str
    subject: str
    description: str | None
    due_date: date
    max_score: float
    file_url: str | None
    submitted_at: datetime | None
    score: float | None
    feedback: str | None


class ParentFinanceItem(BaseModel):
    id: uuid.UUID
    amount_kobo: int
    payment_type: str
    payment_method: str
    academic_session: str
    term: str
    status: str
    reference: str
    paid_at: datetime | None
    created_at: datetime


class _Page(BaseModel):
    total: int
    page: int
    per_page: int
    total_pages: int


class PaginatedParentAttendance(_Page):
    items: list[ParentAttendanceItem]
    present_count: int
    absent_count: int
    late_count: int
    excused_count: int
    attendance_rate: float


class PaginatedParentResults(_Page):
    items: list[ParentResultItem]


class PaginatedParentAssignments(_Page):
    items: list[ParentAssignmentItem]


class PaginatedParentFinance(_Page):
    items: list[ParentFinanceItem]

File: frontend/types/parentPortal.ts
export interface ParentChildPermissions {
  attendance: boolean;
  results: boolean;
  assignments: boolean;
  finance: boolean;
  messaging: boolean;
  pickup: boolean;
}

export interface ParentChild {
  id: string;
  admission_number: string;
  first_name: string;
  middle_name?: string | null;
  last_name: string;
  date_of_birth: string;
  gender: string;
  photo_url?: string | null;
  class_id?: string | null;
  class_name?: string | null;
  class_level?: string | null;
  academic_session?: string | null;
  permissions: ParentChildPermissions;
}

export interface ParentDashboard {
  children: ParentChild[];
  child_count: number;
  attendance_records: number;
  present_records: number;
  attendance_rate: number;
  approved_results: number;
  upcoming_assignments: number;
}

export interface ParentChildrenPage {
  items: ParentChild[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface ParentProfile {
  guardian_id: string;
  name: string;
  email: string;
  phone?: string | null;
  address?: string | null;
  occupation?: string | null;
  preferred_contact_channel: 'in_app' | 'email' | 'sms' | 'whatsapp';
}

export interface ParentAttendanceItem { id: string; date: string; status: 'present' | 'absent' | 'late' | 'excused'; note?: string | null }
export interface ParentAttendancePage {
  items: ParentAttendanceItem[]; total: number; page: number; per_page: number; total_pages: number;
  present_count: number; absent_count: number; late_count: number; excused_count: number; attendance_rate: number;
}
export interface ParentResultItem { id: string; subject: string; academic_session: string; term: string; ca_score?: number | null; exam_score?: number | null; total_score?: number | null; grade?: string | null; teacher_comment?: string | null }
export interface ParentResultsPage { items: ParentResultItem[]; total: number; page: number; per_page: number; total_pages: number }
export interface ParentAssignmentItem { id: string; title: string; subject: string; description?: string | null; due_date: string; max_score: number; file_url?: string | null; submitted_at?: string | null; score?: number | null; feedback?: string | null }
export interface ParentAssignmentsPage { items: ParentAssignmentItem[]; total: number; page: number; per_page: number; total_pages: number }
export interface ParentFinanceItem { id: string; amount_kobo: number; payment_type: string; payment_method: string; academic_session: string; term: string; status: string; reference: string; paid_at?: string | null; created_at: string }
export interface ParentFinancePage { items: ParentFinanceItem[]; total: number; page: number; per_page: number; total_pages: number }
